MaxText/layers/attentions.py

```diff
@@ -260,22 +241,174 @@ def key_rotary(self, key: Array, inputs_positions: Array):
+      mask = mask[:, None, None,:, :]
+
+    _, q_seq_len, _, _ = query.shape
+    _, kv_seq_len, _, _ = key.shape
+    mask_shape = (q_seq_len, kv_seq_len)
+    row_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 0)
+    col_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 1)
+    causal_mask = (col_ids <= row_ids)[None, None, None, :, :]
+
+    mask = causal_mask if mask is None else jnp.logical_and(mask, causal_mask)
+    return jnp.where(mask, 0.0, DEFAULT_MASK_VALUE) if mask is not None else None
+  
+  def apply_attention(self,
       query: Array,
       key: Array,
       value: Array,
       decoder_segment_ids: Array | None,
-      attention_bias: Array | None,
       dropout_rng: PRNGKey | None,
       deterministic: bool,
-      decode: bool = False,
+      decode: bool = False) -> Array:
+    
+    if self.attention_kernel == "dot_product":
+      return self.apply_attention_dot(query, key, value, decoder_segment_ids, dropout_rng, deterministic, decode)
+    elif self.attention_kernel == 'flash':
+      if decode:
+        raise ValueError("""Decode not supported with flash attention.
+                            Use `dot_product` instead.""")
+      return self.tpu_flash_attention(query, key, value, decoder_segment_ids)
+    elif self.attention_kernel == 'gpu_flash' or self.attention_kernel == 'gpu_flash_triton':
+      if decode:
+        raise ValueError("""Decode not supported with flash attention.
+                            Use `dot_product` instead.""")
+      return self.gpu_flash_attention(query, key, value)
+    else:
+      raise ValueError('Unexpected attention type.')
+
+  def tpu_flash_attention(
+    self,
+    query: Array,
+    key: Array,
+    value: Array,
+    decoder_segment_ids: Array | None) -> Array:
+    """TPU Flash Attention."""
+    # Transpose to ('batch', 'heads', 'length', 'kv')
+    query = jax.numpy.transpose(query, axes=(0, 2, 1, 3))
+    key = jax.numpy.transpose(key, axes=(0, 2, 1, 3))
+    value = jax.numpy.transpose(value, axes=(0, 2, 1, 3))
+    if not(query.shape[1] == key.shape[1] == value.shape[1]):
+      raise ValueError(f"The flash attention kernel requires Q, K and V to have the same number of heads"
+                       "{query.shape=} {key.shape=}, {value.shape=}")
+
+    if decoder_segment_ids is not None:
+      decoder_segment_ids = tpu_flash_attention.SegmentIds(
+          decoder_segment_ids, decoder_segment_ids
+      )
+    axis_names = nn.logical_to_mesh_axes(self.flash_axis_names)
+    segment_axis_names = nn.logical_to_mesh_axes(
+        (BATCH, 'activation_length_no_heads')
+    )
+
+    @functools.partial(
+        shard_map,
+        mesh=self.mesh,
+        in_specs=(
+            axis_names,
+            axis_names,
+            axis_names,
+            segment_axis_names,
+        ),
+        out_specs=axis_names,
+        check_rep=False,
+    )
+    def wrap_flash_attention(query, key, value, decoder_segment_ids):
```

Comment:
No docstring.

Reply:
Legitimate feedback but I'm just trying to move things around here! Thanks for calling it out.